r/R/metadata.R

```diff
@@ -133,24 +133,24 @@ remove_attributes <- function(x) {
 }
 
 arrow_attributes <- function(x, only_top_level = FALSE) {
+
+  att <- attributes(x)
+  removed_attributes <- remove_attributes(x)
+
   if (inherits(x, "grouped_df")) {
     # Keep only the group var names, not the rest of the cached data that dplyr
     # uses, which may be large
     if (requireNamespace("dplyr", quietly = TRUE)) {
       gv <- dplyr::group_vars(x)
       x <- dplyr::ungroup(x)
       # ungroup() first, then set attribute, bc ungroup() would erase it
-      attr(x, ".group_vars") <- gv
-    } else {
-      # Regardless, we shouldn't keep groups around
-      attr(x, "groups") <- NULL
+      att[[".group_vars"]] <- gv
+      removed_attributes <- c(removed_attributes, "groups", "class")
```

Comment:
"class" should already be in `removed_attributes` right? Or am I misunderstanding the flow up there?

Reply:
When `remove_attributes()` is called on `x` it hits this block of code:
`
if (identical(class(x), c("tbl_df", "tbl", "data.frame"))) {
    removed_attributes <- c("class", "row.names", "names")
  } else if (inherits(x, "data.frame")) {
    removed_attributes <- c("row.names", "names")
`
As it's a `grouped_df` it fails the first condition and hits the second.  That doesn't include "class" for reasons I don't fully understand but its addition makes other tests fail and changes the returned object type.  I've tried a few other approaches but they make other tests fail, and I'm going down a bit of a rabbit hole.

Reply:
Aaaah, I missed that `identical()` bit there the first time I was reading. I presume one of the branches you tried was something like `all(c("tbl_df", "tbl", "data.frame") %in% class(x)))` and that didn't work?

I'm fine to keep it here if that doesn't work (/ causes other issues).

Reply:
Yeah - there are checks for object class in other tests and those began to fail.  How about we leave it as is for now, but fix later if other issues arise? I wanted to find a simple solution but it's more complex than I thought.

Reply:
Sounds good!